### ISSUE TITLE
[Modding] Client to server communication

### ISSUE THREAD

**Issue by JoelTroch** [ValveSoftware/halflife#1613]
**TL;DR**
To make a server to client communication, we use messages like this one :
```cpp
MESSAGE_BEGIN( MSG_ALL, gmsgMyMessage );
    WRITE_BYTE( 1 );
    WRITE_STRING( "Hello World !" );
MESSAGE_END();
```
How to do the reverse way (client -> server) without using client commands like TFC and Cry of Fear did ?

**Detailed version**
Few years ago on the HL coders mailing list, people asked how to do client to server communication, TFC and the popular Cry of Fear standalone Goldsrc game relied on client commands (``HOOK_COMMAND( "myclientcommand", MyClientCommand );``) that the server is calling to achieve this (``CLIENT_COMMAND( edict(), "myclientcommand" );``). However, there are problems with this technique :

1) I don't find this way very elegant.
2) If the player knows/discover these commands, it could be used as cheating.

This could be annoying for people who would like to make some gameplay stuff like inventory, door's keypad and so on. So the question is how can we make clean/proper client to server communication without having to use client commands ?

Here's the link to the original HL coders mail : https://www.mail-archive.com/hlcoders%40list.valvesoftware.com/msg31740.html

**Comment by tschumann** [comment]
So you're looking for a way to set player data on the client and be able to read that on the server?

**Comment by JoelTroch** [comment]
Yes

**Comment by tschumann** [comment]
It looks like client command are probably the most elegant way. You might be able to store values in the user info key value store (and have cvars as proxies to those values using the FCVAR_USERINFO flag) but that is much less elegant I think.
Off the top of my head I don't think there are any other (obvious/useful) ways to send a value to the server.

**Comment by tschumann** [comment]
Did you ever work anything out?
